deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/EncodingHandler.java

```diff
@@ -56,14 +77,57 @@ public void initialize(@NonNull GradientsAccumulator accumulator) {
 
     public INDArray encodeUpdates(INDArray updates) {
         // special op should be called here for encoding
+        if (bitmapMode.get() == null) {
+            bitmapMode.set(new AtomicBoolean(true));
+            currentThreshold.set(new AtomicDouble(threshold));
+            iterations.set(new AtomicLong(0));
+            lastStep.set(new AtomicLong(0));
+        }
 
+        iterations.get().incrementAndGet();
 
         if (boundary != null && atomicBoundary.get() < 0)
             atomicBoundary.compareAndSet(-1, (int) (updates.lengthLong() * boundary));
 
-        //return compressor.compress(updates);
-        INDArray encoded = Nd4j.getExecutioner().thresholdEncode(updates, threshold,
-                        boundary == null ? null : atomicBoundary.get());
+        INDArray encoded =  null;
+
+        if (!bitmapMode.get().get()) {
+            // if shakeFrequency hits here, we'll use bitmap encoding for one round for 1/3 of current threshold
+            if (shakeFrequency != 0 && iterations.get().get() % shakeFrequency == 0) {
+                DataBuffer buffer = Nd4j.getDataBufferFactory().createInt(updates.lengthLong() / 16 + 5);
+                encoded = Nd4j.createArrayFromShapeBuffer(buffer, updates.shapeInfoDataBuffer());
+
+                Nd4j.getExecutioner().bitmapEncode(updates, encoded, currentThreshold.get().get() / 3);
+            } else {
+                // otherwise (probably most often - we go for sparse
+                encoded = Nd4j.getExecutioner().thresholdEncode(updates, currentThreshold.get().get(), boundary == null ? null : atomicBoundary.get());
+                double encLen = encoded.data().getInt(0);
+                double encodingRatio = encLen * 100.0 / updates.length();
+
+                // if updates are too dense - we fallback to bitmap encoding
+                if (encLen > (updates.lengthLong() / 16 + 5))
+                    bitmapMode.get().set(true);
+
+
+                // after encoding is finished, and updates are sparse enough - let's step down a bit
+                // and we don't step down too early, so we wait for 50 iterations at least to step down
+                if (minThreshold <= currentThreshold.get().get() && minThreshold < currentThreshold.get().get() - thresholdStep && iterations.get().get() > lastStep.get().get() + 50 && encodingRatio < 0.05) {
```

Comment:
Hardcoded threshold/constants here - seems to be reasonable, but should this perhaps be configurable?